tls/s2n_connection.c

```diff
@@ -1377,3 +1384,94 @@ int s2n_connection_set_keyshare_by_name_for_testing(struct s2n_connection *conn,
+        sk_X509_pop_free(*cert_chain, X509_free);
+        *cert_chain = NULL;
+    }
+    return S2N_SUCCESS;
+}
+
+static int s2n_cert_chain_wipe(struct s2n_cert_chain *cert_chain)
+{
+    /* Walk the chain and free the certs/nodes allocated prior to failure */
+    if (cert_chain) {
+        struct s2n_cert *node = cert_chain->head;
+        while (node) {
+            /* Free the cert */
+            POSIX_GUARD(s2n_free(&node->raw));
+            /* update head so it won't point to freed memory */
+            cert_chain->head = node->next;
+            /* Free the node */
+            POSIX_GUARD(s2n_free_object((uint8_t **)&node, sizeof(struct s2n_cert)));
+            node = cert_chain->head;
+        }
+    }
+
+    return S2N_SUCCESS;
+}
+
+int s2n_connection_get_peer_cert_chain(const struct s2n_connection *conn, struct s2n_cert_chain_and_key *cert_chain_and_key)
+{
+    POSIX_ENSURE_REF(conn);
+    POSIX_ENSURE_REF(cert_chain_and_key);
+
+    struct s2n_cert_chain *cert_chain = cert_chain_and_key->cert_chain;
+    POSIX_ENSURE_REF(cert_chain);
+    struct s2n_cert **insert = &cert_chain->head;
+    POSIX_ENSURE(*insert == NULL, S2N_ERR_INVALID_ARGUMENT);
+
+    const struct s2n_x509_validator *validator = &conn->x509_validator;
+    POSIX_ENSURE_REF(validator);
+    POSIX_ENSURE(validator->state == VALIDATED, S2N_ERR_CERT_NOT_VALIDATED);
+
+    /* X509_STORE_CTX_get1_chain() returns a validated cert chain if a previous call to X509_verify_cert() was successful.
+     * X509_STORE_CTX_get0_chain() is a better API because it doesn't return a copy. But it's not available for Openssl 1.0.2.
+     * See the comments here:
+     * https://www.openssl.org/docs/man1.0.2/man3/X509_STORE_CTX_get1_chain.html
+     */
+    DEFER_CLEANUP(STACK_OF(X509) *cert_chain_validated = X509_STORE_CTX_get1_chain(validator->store_ctx),
+                  s2n_sk_X509_pop_free);
+    POSIX_ENSURE_REF(cert_chain_validated);
+
+    for (size_t cert_idx = 0; cert_idx < sk_X509_num(cert_chain_validated); cert_idx++) {
+        X509 *cert = sk_X509_value(cert_chain_validated, cert_idx);
+        POSIX_ENSURE_REF(cert);
+        DEFER_CLEANUP(uint8_t *cert_data = NULL, s2n_openssl_free);
+        int cert_size = i2d_X509(cert, &cert_data);
+        POSIX_ENSURE_GT(cert_size, 0);
+
+        struct s2n_blob mem = { 0 };
+        if (s2n_alloc(&mem, sizeof(struct s2n_cert)) < S2N_SUCCESS) {
+            goto cleanup;
+        }
+
+        struct s2n_cert *new_node = (struct s2n_cert *)(void *)mem.data;
+        if (s2n_alloc(&new_node->raw, cert_size) < S2N_SUCCESS) {
+            goto cleanup;            
+        }
+        POSIX_CHECKED_MEMCPY(new_node->raw.data, cert_data, cert_size);
+
+        new_node->next = NULL;
+        *insert = new_node;
+        insert = &new_node->next;
+
+    }
+
+    return S2N_SUCCESS;
+    
+    cleanup:
+        POSIX_GUARD(s2n_cert_chain_wipe(cert_chain));
+        POSIX_BAIL(S2N_ERR_INVALID_CERT_STATE);
```

Comment:
I would strongly suggest you use DEFER_CLEANUP + ZERO_TO_DISABLE_DEFER_CLEANUP here instead. What if your method fails somewhere other than where you're explicitly checking for alloc failures? There are other GUARDs in that loop.

Alternatively, wrap this method in another and always wipe the chain on a failure.

Reply:
Actually there aren't any other GUARD'S in the scope/loop and it would not fail other than these calls. Nice to know ZERO_TO_DISABLE_DEFER_CLEANUP exists, I will use it to keep it cleaner and prevent goto statements as I am not happy with the goto either.